[libc++] Implement LWG3545: std::pointer_traits should be SFINAE-friendly.

zetafunction opened the pull request:
See https://wg21.link/LWG3545 for background and details.

Differential Revision: https://reviews.llvm.org/D158922

One question I have is if this needs to be limited to C++23 and above. For full disclosure, it would be more useful to me if it were not, and it seems like it'd be OK to allow this to leak through to previous C++ versions, since those programs wouldn't have compiled anyway.

zetafunction committed:
[libc++] Implement LWG3545: std::pointer_traits should be SFINAE-friendly.

See https://wg21.link/LWG3545 for background and details.

Differential Revision: https://reviews.llvm.org/D158922

EricWF commented on libcxx/test/std/utilities/memory/pointer.conversion/to_address_without_pointer_traits.pass.cpp:
```diff
@@ -0,0 +1,48 @@
+
+  int* ptr;
+};
+
```
You need to add some tests for the actual SFINAE friendliness of the type. You'll need some sort of metafunction that allows you to test code in a SFINAE context. For example:

```
template <template <class...> class Templ, class Ignore, class... Args>
struct is_valid_expansion_impl
{
    typedef std::false_type type;
};

template <template <class...> class Templ, class... Args>
struct is_valid_expansion_impl<Templ, decltype((void) Templ<Args...>{}, 0), Args...>
{
    typedef std::true_type type;
};

template <template <class...> class Templ, class... Args>
using is_valid_expansion = typename is_valid_expansion_impl<Templ, int, Args...>::type;


template <class Ptr>
using TestToAddressCall = decltype(std::to_address(std::declval<Ptr>()))

static_assert(is_valid_expansion<TestToAddress, IntPtr>>::value);
static_assert(not is_valid_expansion<TestToAddress, std::string>>::value); 
```


At least that's the least ugly way I can think to test to_address without using concepts

zetafunction replied:
OK, that's very clever. I added the tests you suggested, as well as some direct tests of pointer_traits themselves.

It seems like rebind and pointer_to aren't SFINAE-friendly either, but I was able to test that pointer_traits for a non-pointer-like type is empty at least.

zetafunction replied:
By the way, do we generally follow LLVM naming conventions in libc++ tests? I'm not sure if it's better to name traits `like_this` or `LikeThat`.

ldionne replied:
I think the snake_case you have right now is reasonable.

EricWF reviewed (changes requested):
The implementation looks good, but the tests are lacking.

We need to add tests that substitution actually fails without causing a hard compile error.

zetafunction commented on libcxx/test/std/utilities/memory/pointer.conversion/to_address.verify.cpp:
```diff
@@ -0,0 +1,23 @@
+struct NotPtr {};
+
+void test() {
+  (void)std::to_address(NotPtr()); // expected-error@*:* {{no matching function for call to 'to_address'}}
```
This test was mostly making sure that `std::to_address()` on non-pointer-like types still fails. If you don't think this is useful, happy to remove this.

ldionne replied:
I would suggest putting this test in `libcxx/test/std/utilities/memory/pointer.conversion/to_address.pass.cpp` and implementing it like this instead:

```
template <class T>
concept has_to_address = requires(T t) {
  std::to_address(t);
};

static_assert(!has_to_address<NotPtr>);
```

This way, you are actually checking that `to_address` will SFINAE-away if it is passed something that's not a pointer. With your current `.verify.cpp` test, the test would also pass if there was a hard error while trying to instantiate this and it wasn't SFINAE friendly. Does this make sense?

Note that you'll need to change the code I put above to avoid requiring concepts, but I was trying to express the idea here.

ldionne replied:
I don't think this test file is relevant anymore, right?

zetafunction replied:
Removed.

ldionne commented on libcxx/test/std/utilities/memory/pointer.conversion/pointer_traits.pass.cpp:
```diff
@@ -0,0 +1,176 @@
+//===----------------------------------------------------------------------===//
```
These tests look really nice, but we should merge them with the ones in `libcxx/test/std/utilities/memory/pointer.traits/*` (for example `libcxx/test/std/utilities/memory/pointer.traits/difference_type.pass.cpp`).

I think it might make sense to consolidate them in a single `libcxx/test/std/utilities/memory/pointer.traits/types.pass.cpp`.

ldionne reviewed (changes requested):
Thanks! This is looking pretty good but I still have a few comments.

zetafunction committed:
Consolidate tests and add better coverage of T* specialization.

zetafunction committed:
Test pre-C++20 pointer_traits.

ldionne commented on libcxx/test/std/utilities/memory/pointer.traits/types.pass.cpp:
```diff
@@ -0,0 +1,236 @@
+  element_type* ptr;
+};
+
+constexpr bool test() {
```
You could use `static_assert` instead of `assert` in most places in this file!

zetafunction replied:
I'm mostly emulating the style of previous tests (primarily test/std/utilities/memory/pointer.conversion/to_address.pass.cpp). I had assumed this was intentional to test the constexpr vs non-constexpr paths in the compiler, but I am happy to migrate everything to static_assert.

ldionne commented on libcxx/test/std/utilities/memory/pointer.traits/types.pass.cpp:
```diff
@@ -0,0 +1,236 @@
+    assert(HasDifferenceType<Ptr>::value);
+    ASSERT_SAME_TYPE(typename std::pointer_traits<Ptr>::difference_type, ptrdiff_t);
+
+    // TODO: Consider supporting SFINAE testing of std::pointer_traits<Ptr>.
```
Can you expand on that? Isn't it what this patch is about?

zetafunction replied:
Done. I think LWG3545 leaves it unclear if rebind and pointer_to should be SFINAE-able. It works with libstdc++, but not with libc++ currently.

zetafunction committed:
Support pointer_traits tests through C++14.

zetafunction committed:
Support pointer_traits tests through C++11.

zetafunction committed:
Remove unneeded test.

zetafunction committed:
Clarify comment about SFINAE testing of pointer_traits members.

zetafunction committed:
Support C++03 in this test as well.

The standard doesn't define pointer_traits until C++11, but there are
some existing tests that seem to test in C++03 mode... so try to
preserve that behavior in the updated tests.

ldionne commented on libcxx/test/std/utilities/memory/pointer.traits/types.pass.cpp:
```diff
@@ -0,0 +1,325 @@
+  element_type* ptr;
+};
+
+#if TEST_STD_VER >= 14
+constexpr
+#endif
+    bool
+    test() {
```
Let's use `TEST_CONSTEXPR_CXX14` instead.

ldionne commented on libcxx/test/std/utilities/memory/pointer.traits/types.pass.cpp:
```diff
@@ -0,0 +1,325 @@
+    test() {
+  {
+    using Ptr = NotAPtr;
+    assert(!HasElementType<Ptr>::value);
+    assert(!HasPointerType<Ptr>::value);
+    assert(!HasDifferenceType<Ptr>::value);
+    assert((!HasRebind<Ptr, long>::value));
+    assert(!HasPointerTo<Ptr>::value);
```
```suggestion
    static_assert(!HasElementType<Ptr>::value, "");
    static_assert(!HasPointerType<Ptr>::value, "");
    static_assert(!HasDifferenceType<Ptr>::value, "");
    static_assert(!HasRebind<Ptr, long>::value, "");
    static_assert(!HasPointerTo<Ptr>::value, "");
```

zetafunction replied:
Done.

ldionne commented on libcxx/test/std/utilities/memory/pointer.traits/types.pass.cpp:
```diff
@@ -0,0 +1,325 @@
+  {
+    using Ptr = int*;
+
+    assert(HasElementType<Ptr>::value);
+    ASSERT_SAME_TYPE(typename std::pointer_traits<Ptr>::element_type, int);
+
+    assert(HasPointerType<Ptr>::value);
+    ASSERT_SAME_TYPE(typename std::pointer_traits<Ptr>::pointer, Ptr);
+
+    assert(HasDifferenceType<Ptr>::value);
+    ASSERT_SAME_TYPE(typename std::pointer_traits<Ptr>::difference_type, ptrdiff_t);
+
+    assert((HasRebind<Ptr, long>::value));
```
```suggestion
    static_assert(HasElementType<Ptr>::value, "");
    ASSERT_SAME_TYPE(typename std::pointer_traits<Ptr>::element_type, int);

    static_assert(HasPointerType<Ptr>::value, "");
    ASSERT_SAME_TYPE(typename std::pointer_traits<Ptr>::pointer, Ptr);

    static_assert(HasDifferenceType<Ptr>::value, "");
    ASSERT_SAME_TYPE(typename std::pointer_traits<Ptr>::difference_type, ptrdiff_t);

    static_assert(HasRebind<Ptr, long>::value, "");
```

zetafunction replied:
Done.

ldionne commented on libcxx/test/std/utilities/memory/pointer.traits/types.pass.cpp:
```diff
@@ -0,0 +1,325 @@
+    ASSERT_SAME_TYPE(typename std::pointer_traits<Ptr>::rebind<long>::other, long*);
+#endif
+
+    assert(HasPointerTo<Ptr>::value);
+    int variable = 0;
+    ASSERT_SAME_TYPE(decltype(std::pointer_traits<Ptr>::pointer_to(variable)), Ptr);
+#if TEST_STD_VER >= 17
+    if constexpr (std::__libcpp_is_constant_evaluated() && TEST_STD_VER >= 20) {
+      assert(std::pointer_traits<Ptr>::pointer_to(variable) == &variable);
+    }
+#endif
```
```suggestion
    static_assert(HasPointerTo<Ptr>::value, "");
    int variable = 0;
    ASSERT_SAME_TYPE(decltype(std::pointer_traits<Ptr>::pointer_to(variable)), Ptr);
    assert(std::pointer_traits<Ptr>::pointer_to(variable) == &variable);
```

zetafunction replied:
Done.

ldionne commented on libcxx/test/std/utilities/memory/pointer.traits/types.pass.cpp:
```diff
@@ -0,0 +1,325 @@
+    assert(HasDifferenceType<Ptr>::value);
+    ASSERT_SAME_TYPE(typename std::pointer_traits<Ptr>::difference_type, ptrdiff_t);
+
+    // TODO: Maybe support SFINAE testing of std::pointer_traits<Ptr>::rebind
```
I don't understand this comment.

ldionne commented on libcxx/test/std/utilities/memory/pointer.traits/types.pass.cpp:
```diff
@@ -0,0 +1,325 @@
+    ASSERT_SAME_TYPE(typename std::pointer_traits<Ptr>::rebind<long>::other, long*);
+#endif
+
+    assert(HasPointerTo<Ptr>::value);
```
The test for `pointer_to` shouldn't be in `types.compile.pass.cpp` since it's not a compile-time only test.

zetafunction replied:
If I change it to using static_assert (which I did), is it OK to leave it here?

ldionne commented on libcxx/test/std/utilities/memory/pointer.traits/types.pass.cpp:
```diff
@@ -0,0 +1,325 @@
+//===----------------------------------------------------------------------===//
```
Rename this test to `types.compile.pass.cpp`.

zetafunction replied:
Done.

ldionne commented:
@zetafunction I added some suggestions to try and capture what I want here, feel free to reach out on Discord if you want, sometimes it's easier to transmit the exact intent when chatting live.

zetafunction committed:
Convert everything to use static_assert.
